### PR TITLE
fix: improve interaction between nextjs, yarn 3 and custom output paths

### PR DESCRIPTION
**Pull request by Schmavery**
I've been trying to use prisma with yarn 3 which doesn't have a standard node_modules. I used a [custom output path](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/generating-prisma-client#using-a-custom-output-path) to create the client in my ./src folder instead as suggested in a couple of related issues.

This seemed to work but then I got error at runtime similar to issues like #12823, #9435 and #12921 . I did some digging and realized that the client is failing to find the `schema.prisma` at runtime due to the findSync search being depth first. This makes it quite dependent on the order of the contents of the root directory which might explain why some people seem to "randomly" run into the problem in several different configurations?

In my case, it was picking up the `./.yarn/unplugged/prisma-npm-<version>-<hash>/prisma-client` folder before the `./src/prisma/prisma-client`, the first of which didn't have the prisma-schema. One option would have been to hardcode a check that ignores the ".yarn" folder, but it seems in general that it would be less surprising behavior to pick the match closest to the root of the search.

I have been using this as a patch locally but wanted to share in case it is useful to anyone else. I tried to match the existing style where possible but feel free to tweak it.

### CODE REVIEW

**Comment by aqrln** [comment]
Hi @Schmavery and thank you for the PR! As we're pretty busy at the moment, I apologize if there's any delay in reviewing, but we'll try to review it as soon as possible. Meanwhile, would you mind fixing the build errors? Thank you again!

```
Error: packages/client build: src/runtime/utils/find.ts(94,15): error TS2339: Property 'item' does not exist on type '{ item: Dirent; dir: string; } | undefined'.
Error: packages/client build: src/runtime/utils/find.ts(94,21): error TS2339: Property 'dir' does not exist on type '{ item: Dirent; dir: string; } | undefined'.
Error: packages/client build: src/runtime/utils/find.ts(125,25): error TS2345: Argument of type '"d" | "f" | "l" | undefined' is not assignable to parameter of type '"d" | "l"'.
```

**Comment by Jolg42** [comment]
Note: we probably need to look at this, not only from a yarn 3 point of view (I mean see how it affects, package managers, lik pnpm, and workspaces for example)

@Schmavery we will need tests for this yarn 3 issue that motivated your issue.
Did you provide a minimal reproduction somewhere? It would be great to have one for your case here.

**Comment by Schmavery** [comment]
Hey all, sorry for the build error, will fix those.

Was just working on a minimal repro and realized there was more going on than I expected. Basically the fact that I'm using nextjs seems to be breaking the normal __dirname-based resolution (otherwise the findSync code never even gets called and everything seems to work fine even with yarn 3 😄 ).

Will share the actual repro soon, thanks for taking a look so quickly!

**Comment by Schmavery** [comment]
Here is the repro for now:
https://github.com/Schmavery/prisma-findsync-repro

Will look into setting up the proper prisma contribution environment and fixing the errors when I get a chance.

**Comment by Schmavery** [comment]
@aqrln I fixed the type error and now most of the checks appear to be passing. I admit I'm not sure what the actual errors are in the 2 failing tests. Happy to try and fix but I could use another pointer when you get some time.

**Comment by janpio** [comment]
(Restarted both failing test runs for the case this was flakey)

File: packages/client/src/runtime/utils/find.ts [hunk 1]
```diff
@@ -68,36 +71,45 @@ export function findSync(
...
     // we check that the root is a directory
     if (direntToType(statSync(realRoot)) !== 'd') {
       return found
     }
 
-    // we list the items in the current root
-    const items = readdirSync(root, { withFileTypes: true })
+    // we list the items in the search root
+    const items = readdirSync(realRoot, { withFileTypes: true })
+    const stack = [...items.map((i) => ({ item: i, dir: root }))]
```
Comment (aqrln) [inline packages/client/src/runtime/utils/find.ts:81]: nit: technically it's a queue, not a stack, a stack would lead to depth-first search

Comment (Schmavery) [reply packages/client/src/runtime/utils/find.ts]: hilarious, fixed

**Pull request closed without merging**